### ISSUE TITLE
Basic HTTP Auth header and more can be logged when HTTP client connection runs into an unhandled exception

### ISSUE THREAD

**Issue by cdris** [rabbitmq/rabbitmq-management#474]
Found the following (edited) error in the logs for our heavily bogged down rabbitmq cluster:

```
=ERROR REPORT==== 7-Sep-2017::18:42:50 ===
Ranch listener rabbit_web_dispatch_sup_15671 had connection process started with cowboy_protocol:start_link/4 at <0.32492.0> exit with reason: 
[
{reason,
    {timeout,
        {gen_server,call,[<0.3074.0>,
            {fetch,#Fun<rabbit_mgmt_db.22.104142656>,[[…list of queue info…]]},60000]}}},
{mfa, {rabbit_mgmt_wm_queues,to_json,2}},
{stacktrace,[
    {gen_server,call,3,[
        {file,"gen_server.erl"},
        {line,212}]},
    {rabbit_mgmt_db,submit_cached,4,[
        {file,"src/rabbit_mgmt_db.erl"},
        {line,712}]},
    {rabbit_mgmt_util,augment,2,[
        {file,"src/rabbit_mgmt_util.erl"},
        {line,392}]},
    {rabbit_mgmt_util,run_augmentation,2,[
        {file,"src/rabbit_mgmt_util.erl"},
        {line,369}]},
    {rabbit_mgmt_util,augment_resources0,6,[
        {file,"src/rabbit_mgmt_util.erl"},
        {line,358}]},
    {rabbit_mgmt_util,with_valid_pagination,3,[
        {file,"src/rabbit_mgmt_util.erl"},
        {line,282}]},
    {rabbit_mgmt_wm_queues,to_json,2,[
        {file,"src/rabbit_mgmt_wm_queues.erl"},
        {line,54}]},
    {cowboy_rest,call,3,[
        {file,"src/cowboy_rest.erl"},
        {line,976}]}]}, ...
```

The error message proceeds to include every field from the http header, including the full connection cookie and the base64 encoded basic authorization. It also includes context with the username and plaintext password of the account logged in on the management interface.

Can the password and the authorization header fields be removed from logging?

**Comment by michaelklishin** [comment]
Unhandled exceptions are logged by the runtime. Starting with 3.7 there will be a single log file so by limiting logging severity you should be able to silent most of them (I don't know if it's a good idea, however).

The only real solution to make sure HTTP API authentication headers are never accidentally logged by this plugin or an intermediary (e.g. a proxy) is to switch to OAuth 2 tokens, which has been tried twice ([one](https://github.com/rabbitmq/rabbitmq-auth-backend-oauth2), [two](https://github.com/rabbitmq/rabbitmq-auth-backend-uaa)) but there is no consensus on how it should work because whatever solution we come up with must work for Cloud Foundry. That involves getting a lot of teams in the room and so far hasn't lead to much.

**Comment by michaelklishin** [comment]
@essen is there a way to limit what connection state is logged on process termination? I can't think of any but all of this applies to any Cowboy-based API that may use basic HTTP auth, so worth asking :)

**Comment by michaelklishin** [comment]
According to OTP docs, the only way to hook into unhandled process termination handling is by registering [an error handler](http://erlang.org/doc/man/error_logger.html). This is a node-wide thing that is entirely unaware of the context, e.g. what application (plugin) the process belongs to and what it is.

Disabling all crash report logging for this issue alone would be very unfortunate: it is extremely useful when investigating all kinds of issues. It's quite common that we do not have much other information to work with.

@essen @lukebakken @hairyhum @dumbbell @dcorbacho do you know if is there a way for a specific Erlang app such as this plugin to limit how crash reports are formatted without affecting the rest of the system? In 3.6.x we have a separate log file for HTTP API requests, so we can consider using a separate file for the entire plugin. It still wouldn't really address the fundamental problem, though.

**Comment by essen** [comment]
There is a `secure` process_flag that may or may not be used for this, you'll need to experiment. :-) But that will probably prevent logging a lot more too.

**Comment by lukebakken** [comment]
I suggest we limit permissions on the log directory to `0640` (currently `0644`):

```
root@ubuntu-xenial:~# ls -la /var/log/rabbitmq
total 12
drwxr-xr-x 2 rabbitmq rabbitmq 4096 Sep  9 14:19 .
drwxrwxr-x 8 root     syslog   4096 Sep  9 14:19 ..
-rw-r--r-- 1 rabbitmq rabbitmq 2492 Sep  9 14:19 rabbit@ubuntu-xenial.log
-rw-r--r-- 1 rabbitmq rabbitmq    0 Sep  9 14:19 rabbit@ubuntu-xenial-sasl.log
```

That would limit read access to members of the `rabbitmq` group, which would allow operators to grant access as needed.

**Comment by michaelklishin** [comment]
Changing read permissions will break the tools used to tail logs in order to aggregate them today, so if we do this, it better go in before 3.7.0 RC1 ships.

**Comment by lukebakken** [comment]
Closed via rabbitmq/rabbitmq-server-release#48

**Comment by michaelklishin** [comment]
We applied a quick fix: log directory permissions are no longer world-readable for new installations (or installations when log directory location is overridden). The question of what to do about process state logging when an unexpected process termination occurs remains open: often it's a sole source of information about issues and fine grained filtering is a harder problem than it sounds because there's only one error logger per node.
